### PR TITLE
Add support for transforms in MTGPs

### PR DESCRIPTION
**Pull request by wjmaddox**
This PR adds support for input and outcome transforms for MTGPs as currently implemented in BoTorch.

## Motivation

Helps improve bayesian optimization loops using MTGPs including contextual ones by allowing these types of transforms.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

yes.

## Test Plan

Current unit tests; need to dive into these.

## Related PRs

See #679 as well.

### CODE REVIEW

File: botorch/models/gpytorch.py [hunk 1]
```diff
@@ -610,10 +610,10 @@ def posterior(
...
             raise ValueError("Too many output indices")
         cls_name = self.__class__.__name__
-        if hasattr(self, "outcome_transform"):
-            raise NotImplementedError(
-                f"Outcome transforms currently not supported by {cls_name}"
-            )
+        #if hasattr(self, "outcome_transform"):
+        #    raise NotImplementedError(
+        #        f"Outcome transforms currently not supported by {cls_name}"
+        #    )
```
Comment (Balandat) [inline botorch/models/gpytorch.py:616]:
```suggestion
```

File: botorch/models/transforms/outcome.py [hunk 2]
```diff
@@ -256,11 +256,11 @@ def untransform_posterior(self, posterior: Posterior) -> Posterior:
...
-        if not self._m == posterior.event_shape[-1]:
-            raise RuntimeError(
-                "Incompatible output dimensions encountered for transform "
-                f"{self._m} and posterior {posterior.event_shape[-1]}"
-            )
+#         if not self._m == posterior.event_shape[-1]:
+#             raise RuntimeError(
+#                 "Incompatible output dimensions encountered for transform "
+#                 f"{self._m} and posterior {posterior.event_shape[-1]}"
+#             )
```
Comment (Balandat) [inline botorch/models/transforms/outcome.py:263]: what exactly is going on here? Is this for dealing with the flattened outputs for the HOGP models?

**Comment by Balandat** [comment]
@qingfeng10 (for MTGP), @sdaulton (for transforms)

File: botorch/models/transforms/outcome.py (hunk 2 above)
Comment (wjmaddox) [reply botorch/models/transforms/outcome.py:263]: Not quite, in the Matheron MTGP implementation, the event shape is `n_train + n_test` (plus batching) which is not the transforms `_m`.

File: test/models/test_multitask.py [hunk 3]
```diff
@@ -199,10 +199,7 @@ def test_MultiTaskGP(self):
             with self.assertRaises(RuntimeError):
                 MultiTaskGP(train_X, train_Y, 0, output_tasks=[2])
 
-            # test error if outcome_transform attribute is present
-            model.outcome_transform = None
-            with self.assertRaises(NotImplementedError):
-                model.posterior(test_x)
+            # no need to test that outcome transform error is raised
```
Comment (Balandat) [inline test/models/test_multitask.py:202]:
```suggestion
```

File: test/models/test_multitask.py [hunk 4]
```diff
@@ -199,10 +208,16 @@ def test_MultiTaskGP(self):
...
-                model.posterior(test_x)
+            # test outcome transform
+            if use_octf:
+                # ensure un-transformation is applied
+                tmp_tf = model.outcome_transform
+                del model.outcome_transform
+                p_tf = model.posterior(test_x)
+                model.outcome_transform = tmp_tf
+                expected_var = tmp_tf.untransform_posterior(p_tf).variance
+                print(posterior_f.variance, expected_var)
```
Comment (Balandat) [inline test/models/test_multitask.py:219]:
```suggestion
                tmp_tf = model.outcome_transform
                del model.outcome_transform
                p_utf = model.posterior(test_x)
                model.outcome_transform = tmp_tf
                expected_var = tmp_tf.untransform_posterior(p_utf).variance
```

**Review by Balandat — COMMENTED** [review]
> Not quite, in the Matheron MTGP implementation, the event shape is n_train + n_test (plus batching) which is not the transforms _m.

Hmm not sure I fully understand - shouldn't the `event_shape` be something like `batch_shape x (n_train + n_test) x m` rather than `batch_shape x n_test x m` (the standard case)? Then that check shouldn't be an issue?

File: botorch/models/contextual_multioutput.py [hunk 5]
```diff
@@ -8,6 +8,8 @@
 
 import torch
 from botorch.models.multitask import MultiTaskGP
+from botorch.models.transforms.outcome import OutcomeTransform
+from botorch.models.transforms.input import InputTransform
```
Comment (Balandat) [inline botorch/models/contextual_multioutput.py:12]: Internal linter complaining here. I will add a  µsort step to CI soon so this is more obvious. 

```suggestion
from botorch.models.transforms.input import InputTransform
from botorch.models.transforms.outcome import OutcomeTransform
```

**Comment by Balandat** [comment]
For reasons unbeknownst to me this broke `test/models/transforms/test_input.py::TestInputTransforms::test_warp_transform`. It seems that when running this locally via `pytest`, this reliably fails when doing a full run, but reliably passes when testing this test in isolation. It appears there is some weird caching issues going on?

**Pull request closed without merging**